### PR TITLE
[GLUTEN-6876] Support Spark-352

### LINKED ISSUE #6876: Support Spark 352
### Description

Spark 352 has been release and it contains many important bug fixes.
https://spark.apache.org/releases/spark-release-3-5-2.html

### PR DESCRIPTION
**Pull request by zhouyuan**
## What changes were proposed in this pull request?

Support Spark 352

fixes #6876

## How was this patch tested?
Pass GHA

### CODE REVIEW

File: gluten-substrait/src/main/scala/org/apache/spark/shuffle/sort/ColumnarShuffleManager.scala [hunk 1]
```diff
@@ -107,7 +107,7 @@ class ColumnarShuffleManager(conf: SparkConf) extends ShuffleManager with Loggin
           metrics,
           shuffleExecutorComponents)
       case other: BaseShuffleHandle[K @unchecked, V @unchecked, _] =>
-        new SortShuffleWriter(other, mapId, context, shuffleExecutorComponents)
+        new SortShuffleWriter(other, mapId, context, _, shuffleExecutorComponents)
```
Comment (wForget) [inline gluten-substrait/src/main/scala/org/apache/spark/shuffle/sort/ColumnarShuffleManager.scala:110]: Does this mean that gluten will not guarantee spark minor version compatibility?

Comment (zhouyuan) [reply gluten-substrait/src/main/scala/org/apache/spark/shuffle/sort/ColumnarShuffleManager.scala:110]: Hi @wForget, 
Not guaranteed for all minor versions indeed. Actually it's the same for Spark 32 & 3.3 & 3.4. It's only tested with the latest minor release. 
The issue is mostly due to the shim layer design, and the CI/CD resources are also not enough for so many small versions.

File: shims/spark35/src/main/scala/org/apache/spark/sql/execution/GlutenExplainUtils.scala [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+// 3. remove codegen id
+object GlutenExplainUtils extends AdaptiveSparkPlanHelper {
+  def localIdMap: ThreadLocal[java.util.Map[QueryPlan[_], Int]] = QueryPlan.localIdMap
+  type FallbackInfo = (Int, Map[String, String])
+
+  def addFallbackNodeWithReason(
+      p: SparkPlan,
+      reason: String,
+      fallbackNodeToReason: mutable.HashMap[String, String]): Unit = {
+    p.getTagValue(QueryPlan.OP_ID_TAG).foreach {
```
Comment (xhumanoid) [inline shims/spark35/src/main/scala/org/apache/spark/sql/execution/GlutenExplainUtils.scala:53]: https://issues.apache.org/jira/browse/SPARK-48610

QueryPlan.OP_ID_TAG doesn't exist in spark 3.5.2+ / 4.0.0

Comment (zhouyuan) [reply shims/spark35/src/main/scala/org/apache/spark/sql/execution/GlutenExplainUtils.scala]: Thanks, there's also a minor change on sortshufflewriter, will move these changes to shim layer

File: gluten-substrait/src/main/scala/org/apache/spark/shuffle/sort/ColumnarShuffleManager.scala (hunk 1 above)
Comment (wForget) [reply gluten-substrait/src/main/scala/org/apache/spark/shuffle/sort/ColumnarShuffleManager.scala]: I see, thank you.

**Comment by leoluan2009** [comment]
Spark 3.5.3 is out https://spark.apache.org/news/spark-3-5-3-released.html, should we upgrade to 3.5.3?

**Comment by zhouyuan** [comment]
> Spark 3.5.3 is out https://spark.apache.org/news/spark-3-5-3-released.html, should we upgrade to 3.5.3?

Yes, plan to submit one new patch for Spark-353 after this - actually there's not much change from 352 to 353
here's the discussion before the PRC holiday: 
https://github.com/apache/incubator-gluten/issues/7336

thanks, -yuan

**Comment by zhouyuan** [comment]
Run Gluten Clickhouse CI

**Comment by zhouyuan** [comment]
@baibaichen @zhztheplayer

**Comment by baibaichen** [comment]
let me try performance for ch backend,

**Review by baibaichen — APPROVED** [review]
LGTM

**Pull request merged**